### ISSUE TITLE
statistical test for `samplerz`

### ISSUE THREAD

**Issue by aszepieniec** [aszepieniec/falcon-rust#8]
The function `samplerz` claims to sample integers according to a given distribution. I would be nice to verify this using some statistical test.

**Comment by aszepieniec** [comment]
Hi Conner,

Thanks so much for the effort that went into this project. I certainly welcome it and I doubt I will be the only one.

The hat-tip in my direction is much appreciated and I have no problems being named in this way, which is factually correct. I did not review the maths, and the `README.md` file is correct about that too.

One question I am left with after reading `README.md` and `TESTING.md` is how to test the sampler in `falcon-rust`? `TESTING.md` has instructions for generating test vectors, and testing the resulting test vectors, so I guess the right way is to add a unit test that reads from `berexp.json`? It can't hurt to add a section explaining what to do.

I'm happy to review a PR that adds these test vectors; and if you're willing to write it, please do. But you're also welcome to clap your hands and call it a job well done – from the sound of it, no-one can disagree with that. No obligation.

I'm sure you're aware but just in case – there is a discussion going on in the PQC forum on the topic of implementing falcon without floating point arithmetic: https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/hIZo9htrDww/m/vQlJXevlDgAJ?utm_medium=email&utm_source=footer&pli=1 That thread might contain relevant references.

**Comment by spartan8806** [comment]
Good catch — `TESTING.md` was about whether that repo is internally trustworthy and never said how
anyone else would actually use the vectors. Fixed:

- **README → "Wiring them into your own test suite"** — two routes. If `approx_exp` is reachable,
  compute `z` and check `u_drawn < z` directly, no RNG plumbing needed. If only the Bernoulli step is
  exposed, feed the bytes through a stub RNG and compare the boolean. It also says not to assert
  `z == z_reference`, since that is bit-identity with PQClean rather than the 2⁻⁴⁰ question.
- **`TESTING.md`** now says what it covers and points at the other file.

**Two things I would like your call on.**

**1. The PR — `serde_json`, or no new dependency?** The test is written and passing (20/20, and it
fails correctly when I invert a vector's expected answer; your full suite still 169/169). It lives in
`samplerz.rs`'s test module since `ber_exp` is private.

Reading `berexp.json` at test time needs `serde_json` as a dev-dependency, which the crate does not
have today. The alternative is a checked-in `const` table generated from the JSON — no new
dependency, at the cost of a generation step to keep honest. I built it the way you described; say
which you prefer and I will open it.

**2. A small bug I hit while writing it.** `ber_exp` takes `[u8; 7]`, but

```rust
for (index, i) in (0..64).step_by(8).rev().enumerate()
```

yields eight iterations, so `index` reaches `7` and `random_bytes[7]` is out of bounds. Reaching it
requires the top seven random bytes to equal `z`'s top seven — around 2⁻⁵⁶ — and `sign()` draws from
its own `rng()`, so a caller cannot steer it.

Put precisely: every decision is made from the first seven bytes, and the one case that would need
the eighth panics rather than falling through. So it is a 56-bit comparison plus a rare panic on
exact equality, not a silent truncation.

I do not think it is exploitable, but it is a panic in the signing path and the classification is
yours rather than mine. Happy to include a fix in the same PR or keep it separate — whichever you
prefer.

(Unrelated to the bug: every published vector's first divergence from `z` falls at byte index 4 or 5,
so seven bytes is enough to decide all 20 and the test truncates rather than pads.)

**Thanks for the pqc-forum link — it changed something.** It prompted me to add
`pornin/rust-fn-dsa` as a cross-check, and the result was not what I expected.

Both its backends agree 20/20 and reproduce PQClean's 64-bit `z` bit-exactly, including the integer
emulation — though that one is bit-exact by construction, being a strict IEEE-754 `binary64`
emulation. The reason is that its `EXPM_COEFFS` are byte-identical to PQClean's, all 13, same order,
same cited source (ePrint 2018/1234 / `raykzhao/gaussian`). So are `falcon-rust`'s.

So of the three implementations I checked, all evaluate the same polynomial. I had been treating
"find an independent lineage" as the obvious next step for that repo, and that ran two things
together: Pornin's **is** an independent implementation, but not an independent set of coefficients,
and only the second kind would test the polynomial rather than its evaluation. I have not looked
beyond those three, so I make no claim about the wider ecosystem — but within them, the vectors
measure precision of evaluation and could not detect a wrong polynomial. That is now written into
`cross-check/README.md` instead of being left implicit.

— Conner Webber

**Comment by spartan8806** [comment]
On a personal note — thank you for the pqc-forum link. I didn't know that's where this gets discussed; I've mostly been reading the papers and haven't really found my way into the community yet. Genuinely useful. 

And since it's relevant to why: the falcon-rust report was the first post-quantum finding I had the confidence to actually send to anyone. The back-and-forth on it is a lot of why this repo exists at all. Appreciate you engaging with it the way you did really helped me out.

**Comment by aszepieniec** [comment]
> 1. The PR — serde_json, or no new dependency?

Make it into a dev-dependency. Then it will only be imported for developers running the test suite. Downstream users will not import it, unless they need it for other reasons.

> 2. A small bug

So far my policy has been "anything that causes different input-output behavior warrants a new version.". If the fix does not change the input-output behavior, except for crashes, then I'm tempted to bump the minor version instead of the major version. I would appreciate a separate PR that can be scrutinized separately and that can be cited separately as justification for the next version.

> pqc-forum

I recommend announcing your repository on the forum. I expect you will get much more substantive engagement -- the people who hang out there are more knowledgeable than I am.

**Comment by spartan8806** [comment]
Both done — thanks for the clear steers on each.

`serde_json` as a dev-dependency in #15, and the panic fix is up separately as #16. No input-output
change there, so a minor bump should fit your rule: every decision was already coming from the first
seven bytes, since the eighth iteration could only ever panic rather than return. There's a
regression test that builds the tying draw directly instead of waiting 2⁻⁵⁶ for it, and I timed the
hot path — no measurable difference.

Small thing that amused me: the vectors from #15 caught a bug in my own fix while I was writing it. I
had index 0 mapped to shift 48 instead of 56 and all twenty failed instantly. Better advert for them
than anything I put in the README.

One loose end noted at the bottom of #16 rather than fixed — with seven bytes there's no retry when
every byte ties, where PQClean redraws. It's 2⁻⁵⁶ and fixing it would change RNG consumption, so by
your own rule that's a major bump. Left it for you to weigh rather than deciding it for you.

Your point about not asserting 64-bit equality stuck with me, so I went looking at how wide that
choice actually is. Thirteen independently written implementations across eight languages — PQClean,
the designers' own `falcon.py`, this crate, `pornin/rust-fn-dsa`, QSC, wolfSSL, both Bouncy Castles,
`@noble/post-quantum`, and a few smaller ones — all evaluate the same FACCT polynomial. Only the
storage differs. No independent re-derivation anywhere I could reach.

The reason it's so uniform is that the coefficients are sitting in `tprest/falcon.py` marked "lifted
from FACCT", so following the reference means inheriting them. Which makes declining to assert it the
right call for a second reason: it would freeze a convention rather than a requirement. Written up in
`cross-check/README.md` if you're curious.

Taking your pqc-forum advice too — membership request is in, it's moderated so I'm waiting. I'll
reply into the fixed-point thread rather than start a new one, since the survey lands on what those
vectors should be defined over.

Thanks again for all the review. Genuinely enjoyed this one.

**Comment by spartan8806** [comment]
Both done — thanks for the clear steers on each.

`serde_json` as a dev-dependency in #15, and the panic fix is up separately as #16. No input-output
change there, so a minor bump should fit your rule: every decision was already coming from the first
seven bytes, since the eighth iteration could only ever panic rather than return. There's a
regression test that builds the tying draw directly instead of waiting 2⁻⁵⁶ for it, and I timed the
hot path — no measurable difference.

Small thing that amused me: the vectors from #15 caught a bug in my own fix while I was writing it. I
had index 0 mapped to shift 48 instead of 56 and all twenty failed instantly. Better advert for them
than anything I put in the README.

One loose end noted at the bottom of #16 rather than fixed — with seven bytes there's no retry when
every byte ties, where PQClean redraws. It's 2⁻⁵⁶ and fixing it would change RNG consumption, so by
your own rule that's a major bump. Left it for you to weigh rather than deciding it for you.

Your point about not asserting 64-bit equality stuck with me, so I went looking at how wide that
choice actually is. Thirteen independently written implementations across eight languages — PQClean,
the designers' own `falcon.py`, this crate, `pornin/rust-fn-dsa`, QSC, wolfSSL, both Bouncy Castles,
`@noble/post-quantum`, and a few smaller ones — all evaluate the same FACCT polynomial. Only the
storage differs. No independent re-derivation anywhere I could reach.

The reason it's so uniform is that the coefficients are sitting in `tprest/falcon.py` marked "lifted
from FACCT", so following the reference means inheriting them. Which makes declining to assert it the
right call for a second reason: it would freeze a convention rather than a requirement. Written up in
`cross-check/README.md` if you're curious.

Taking your pqc-forum advice too — membership request is in, it's moderated so I'm waiting. I'll
reply into the fixed-point thread rather than start a new one, since the survey lands on what those
vectors should be defined over.

Thanks again for all the review. Genuinely enjoyed this one.
